src/bio/bio_device.c

```diff
@@ -666,18 +691,114 @@ bio_dev_list(struct bio_xs_context *xs_ctxt, d_list_t *dev_list, int *dev_cnt)
 	return rc;
 }
 
+static void
+led_device_action(void *ctx, struct spdk_pci_device *pci_device)
+{
+	struct led_opts		*opts = ctx;
+	enum spdk_vmd_led_state	 cur_led_state;
+	char			 addr_buf[128];
+	int			 rc;
+
+	if (opts->status != 0)
+		return;
+	if (opts->finished)
+		return;
+
+	if (strcmp(spdk_pci_device_get_type(pci_device), "vmd") != 0) {
+		D_ERROR("Found unexpected non-VMD device type\n");
+		opts->status = -DER_INVAL;
+		return;
+	}
+
+	if (!opts->all_devices) {
+		if (spdk_pci_addr_compare(&opts->pci_addr, &pci_device->addr) != 0)
+			return;
+		opts->finished = true;
+	}
+
+	rc = spdk_pci_addr_fmt(addr_buf, sizeof(addr_buf), &pci_device->addr);
+	if (rc != 0) {
+		D_ERROR("Failed to format VMD's PCI address\n");
+		opts->status = -DER_INVAL;
+		return;
+	}
+
+	/* First check the current state of the VMD LED */
+	rc = spdk_vmd_get_led_state(pci_device, &cur_led_state);
+	if (spdk_unlikely(rc != 0)) {
+		D_ERROR("Failed to retrieve the state of the LED on %s\n", addr_buf);
+		opts->status = -DER_INVAL;
+		return;
+	}
+
+	D_DEBUG(DB_MGMT, "LED on %s: %s\n", addr_buf, g_led_states[cur_led_state]);
+
+	switch (opts->action) {
+	case LED_ACTION_GET:
+		opts->led_state = cur_led_state;
+		return;
+
+	case LED_ACTION_SET:
+		D_DEBUG(DB_MGMT, "Setting VMD device %s LED state to %s\n", addr_buf,
+			g_led_states[opts->led_state]);
+		break;
+
+	case LED_ACTION_RESET:
+		/* If the current state of a device is FAULTY we do not want to reset */
+		if (cur_led_state == SPDK_VMD_LED_STATE_FAULT) {
+			D_DEBUG(DB_MGMT, "ignoring LED reset on %s as state is %s\n",
+				addr_buf, g_led_states[cur_led_state]);
+			opts->led_state = cur_led_state;
+			return;
+		}
+		D_DEBUG(DB_MGMT, "Resetting VMD device %s LED state to %s\n", addr_buf,
+			g_led_states[opts->led_state]);
+		break;
+
+	default:
+		D_ERROR("Unrecognized LED action requested\n");
+		opts->status = -DER_INVAL;
+		return;
+	}
+
+	if (cur_led_state == opts->led_state) {
+		D_DEBUG(DB_MGMT, "VMD device %s LED state already in state %s\n", addr_buf,
+			g_led_states[opts->led_state]);
+		return;
+	}
+
+	/* Set the LED to the new state */
+	rc = spdk_vmd_set_led_state(pci_device, opts->led_state);
+	if (spdk_unlikely(rc != 0)) {
+		D_ERROR("Failed to set the LED state on %s\n", addr_buf);
+		opts->status = -DER_NOSYS;
+		return;
+	}
+
+	rc = spdk_vmd_get_led_state(pci_device, &cur_led_state);
+	if (rc) {
+		D_ERROR("Failed to get the VMD LED state\n");
+		opts->status = -DER_NOSYS;
+		return;
+	}
+
+	/* Verify the correct state is set */
+	if (cur_led_state != opts->led_state) {
+		D_ERROR("Unexpected LED state on %s, want %s got %s\n", addr_buf,
+			g_led_states[opts->led_state], g_led_states[cur_led_state]);
+		opts->status = -DER_INVAL;
+	}
```

Comment:
I don't think these return codes are a good representation of what happened in these failure cases.

We have:
- Requested operation failed.
- Unable to verify the result of the operation.
- The result of the operation is wrong/unexpected.

Rather than trying to shoehorn existing error codes, I'd suggest adding new ones for these cases to daos_errno.h.

Reply:
done